api/src/websocket/handlers/subscribe.ts

```diff
@@ -0,0 +1,259 @@
+				// push the event through the Redis pub/sub
+				// logger.debug(`[ WS ] event ${event} - ${JSON.stringify(message)}`);
+				this.messenger.publish('websocket.event', message as Record<string, any>);
+			});
+		};
+		for (const module of modules) {
+			bindAction(module + '.create', ({ key }: any) => ({ key }));
+			bindAction(module + '.update', ({ keys }: any) => ({ keys }));
+			bindAction(module + '.delete', ({ keys }: any) => ({ keys }));
+		}
+		// listen to the Redis pub/sub and dispatch
+		this.messenger.subscribe('websocket.event', (message: Record<string, any>) => {
+			try {
+				this.dispatch(message as WebSocketEvent);
+			} catch (err) {
+				// dont error on an invalid event from the messenger
+				// logger.error('messenger error - ' + JSON.stringify(err, null, 2));
+			}
+		});
+	}
+	/**
+	 * Register a subscription
+	 * @param subscription
+	 */
+	subscribe(subscription: Subscription) {
+		const { collection } = subscription;
+		if (!this.subscriptions[collection]) {
+			this.subscriptions[collection] = new Set();
+		}
+		this.subscriptions[collection]?.add(subscription);
+	}
+	/**
+	 * Remove a subscription
+	 * @param subscription
+	 */
+	unsubscribe(client: WebSocketClient, uid?: string) {
+		if (uid !== undefined) {
+			const subscription = this.getSubscription(uid);
+			if (subscription && subscription.client === client) {
+				this.subscriptions[subscription.collection]?.delete(subscription);
+			} else {
+				// logger.warn(`Couldn't find subscription with UID="${uid}" for current user`);
+			}
+		} else {
+			for (const key of Object.keys(this.subscriptions)) {
+				const subscriptions = Array.from(this.subscriptions[key] || []);
+				for (let i = subscriptions.length - 1; i >= 0; i--) {
+					const subscription = subscriptions[i];
+					if (!subscription) continue;
+					if (subscription.client === client && (!uid || subscription.uid === uid)) {
+						this.subscriptions[key]?.delete(subscription);
+					}
+				}
+			}
+		}
+	}
+	/**
+	 * Dispatch event to subscriptions
+	 */
+	async dispatch(event: WebSocketEvent) {
+		const subscriptions = this.subscriptions[event.collection] ?? new Set();
+		for (const subscription of subscriptions) {
+			const { client } = subscription;
+			try {
+				client.accountability = await refreshAccountability(client.accountability);
+				const schema = await getSchema();
+				const result =
+					'item' in subscription
+						? await this.getSinglePayload(subscription, client.accountability, schema, event)
+						: await this.getMultiPayload(subscription, client.accountability, schema, event);
+				client.send(fmtMessage('subscription', result, subscription.uid));
+			} catch (err) {
+				handleWebsocketException(client, err, 'subscribe');
+			}
+		}
+	}
+	/**
+	 * Handle incoming (un)subscribe requests
+	 */
+	async onMessage(client: WebSocketClient, message: WebSocketSubscribeMessage) {
+		if (message.type.toUpperCase() === 'SUBSCRIBE') {
```

Comment:
I'm actually not sure about the purpose/reasoning of `trimUpper`, but does this `message.type` need `trimUpper` instead of just `.toUpperCase()`? like how it is being used in:

https://github.com/directus/directus/blob/71304eb0e172245eaf2be3342e82882266129d8a/api/src/websocket/handlers/items.ts#L14

https://github.com/directus/directus/blob/71304eb0e172245eaf2be3342e82882266129d8a/api/src/websocket/controllers/base.ts#L147

Reply:
Aaah yeah the `trimUpper` util is a relic from the past 😬 It was very useful before messages were validated by `zod`.